master/buildbot/changes/bitbucket.py

```diff
@@ -14,184 +14,282 @@
 # Copyright Buildbot Team Members
 
 
-import json
-import time
 from datetime import datetime
 
 from twisted.internet import defer
-from twisted.python import log
-from twisted.web import client
 
 from buildbot import config
 from buildbot.changes import base
 from buildbot.util import bytes2unicode
 from buildbot.util import datetime2epoch
-from buildbot.util import deferredLocked
-from buildbot.util import epoch2datetime
+from buildbot.util import httpclientservice
+from buildbot.util.logger import Logger
+from buildbot.util.state import StateMixin
 
-_UNSPECIFIED = object()
+log = Logger()
 
+BASE_URL = "https://api.bitbucket.org"
+OAUTH_URL = 'https://bitbucket.org/site/oauth2/access_token'
+_GET_TOKEN_DATA = {
+    'grant_type': 'client_credentials'
+}
 
-class BitbucketPullrequestPoller(base.PollingChangeSource):
 
-    compare_attrs = ("owner", "slug", "branch",
-                     "pollInterval", "useTimestamps",
-                     "category", "project", "pollAtLaunch")
+class BitbucketPullrequestPoller(base.ReconfigurablePollingChangeSource,
+                                 StateMixin):
 
-    db_class_name = 'BitbucketPullrequestPoller'
+    compare_attrs = ("owner", "slug", "branches", "pollInterval",
+                     "category", "pollAtLaunch", "name")
+    db_class_name = "BitBuckerPullrequestPoller"
 
-    def __init__(self, owner, slug,
-                 branch=None,
-                 pollInterval=10 * 60,
-                 useTimestamps=True,
-                 category=None,
-                 project='',
-                 pullrequest_filter=True,
-                 encoding=_UNSPECIFIED,
-                 pollAtLaunch=False
-                 ):
+    def __init__(self, owner, slug, **kwargs):
+        name = kwargs.get("name")
+        if not name:
+            kwargs["name"] = "BitbucketPullrequestPoller:" + owner + "/" + slug
+        super(BitbucketPullrequestPoller, self).__init__(owner, slug, **kwargs)
+        self.debug = kwargs.get("debug")
+        self.verify = None
+        self.codebase = kwargs.get('codebase')
+
+    def checkConfig(self,
+                    owner,
+                    slug,
+                    branches=None,
+                    pollInterval=10 * 60,
+                    pollAtLaunch=False,
+                    category='pull',
+                    pullrequest_filter=True,
+                    codebase='',
+                    oauth_key=None,
+                    oauth_secret=None,
+                    repository_type="https",
+                    debug=False,
+                    **kwargs):
+        if repository_type not in ["https", "ssh"]:
+            config.error(
+                "repository_type must be one of {https, svn, git, ssh}")
```

Comment:
i am not sure how the error message corresponds to the condition... please add a comment clarifying the relationship or change the condition and/or the message